packages/libs/error-stack/src/ext/stream.rs

```diff
@@ -0,0 +1,420 @@
+use core::{
+    fmt::{Debug, Display},
+    task::Poll,
+};
+
+use futures_core::Stream;
+use pin_project::pin_project;
+
+use crate::{Context, Report, ResultExt};
+
+// note: called this so as to not conflict with other `StreamExt`s
+/// Extension trait for [`Stream`] to provide contextual information on [`Report`]s.
+///
+/// [`Report`]: crate::Report
+pub trait StreamReportExt: Stream + Sized {
+    /// Adds a new attachment to the [`Report`] inside the [`Result`] when
+    /// calling [`Stream::poll_next`].
+    ///
+    /// Applies [`Report::attach`] to the [`Err`] variant, refer to it for more information.
+    ///
+    /// [`Report`]: crate::Report
+    /// [`Report::attach`]: crate::Report::attach
+    fn attach<A>(self, attachment: A) -> StreamWithAttachment<Self, A>
+    where
+        A: Send + Sync + 'static;
+
+    /// Lazily adds a new attachment to the [`Report`] inside the [`Result`] calling
+    /// [`Stream::poll_next`].
+    ///
+    /// Applies [`Report::attach`] to the [`Err`] variant, refer to it for more information.
+    ///
+    /// [`Report`]: crate::Report
+    /// [`Report::attach`]: crate::Report::attach
+    fn attach_lazy<A, F>(self, attachment: F) -> StreamWithLazyAttachment<Self, F>
+    where
+        A: Send + Sync + 'static,
+        F: FnOnce() -> A;
```

Comment:
We only want `F` to be `Fn` because it will be called multiple times in the stream (same as in #716). This also applies to other usages in this file.